autoscale-groups-nodes-plugin/autoscale-groups-nodes-plugin-core/src/main/java/fish/payara/extensions/autoscale/groups/nodes/NodesScaler.java

```diff
@@ -0,0 +1,405 @@
+            }
+
+            if (nodes.getNode(nodeRef) == null) {
+                throw new CommandValidationException("Node " + nodeRef + " does not appear to exist!");
+            }
+        }
+
+
+    }
+
+    @Override
+    public ActionReport scaleUp(int numberOfNewInstances, ScalingGroup scalingGroup) {
+        ActionReport actionReport = commandRunner.getActionReport("plain");
+        try {
+            validate(numberOfNewInstances, scalingGroup);
+        } catch (CommandValidationException commandValidationException) {
+            actionReport.setMessage("Scale up operation cancelled: an error was encountered during validation");
+            actionReport.setFailureCause(commandValidationException);
+            actionReport.setActionExitCode(ActionReport.ExitCode.FAILURE);
+            return actionReport;
+        }
+
+        try {
+            // Create the instances (we currently fail out if we fail to create a single one)
+            List<String> instanceNames = createInstances(numberOfNewInstances, scalingGroup,
+                    actionReport.addSubActionsReport());
+            // Attempt to start the instances
+            startInstances(instanceNames, actionReport.addSubActionsReport());
+        } catch (CommandException commandException) {
+            actionReport.setFailureCause(commandException);
+            actionReport.setActionExitCode(ActionReport.ExitCode.FAILURE);
+            return actionReport;
+        }
+
+        return actionReport;
+    }
+
+    /**
+     * Creates the requested number of instances using the {@link NodesScalingGroup scaling group} config
+     *
+     * @param numberOfNewInstances The number of instances to create
+     * @param scalingGroup The scaling group we're creating the instances against
+     * @param actionReport The action report we want to add out command outputs to
+     * @return A List containing the names of all created instances
+     * @throws CommandException If there's an error creating any instances.
+     */
+    private List<String> createInstances(int numberOfNewInstances, ScalingGroup scalingGroup,
+            ActionReport actionReport) throws CommandException {
+        List<String> instanceNames = new ArrayList<>();
+
+        // Get the starting balance of instances to nodes
+        DeploymentGroup deploymentGroup = deploymentGroups.getDeploymentGroup(scalingGroup.getDeploymentGroupRef());
+        Map<String, Integer> scalingGroupBalance = getNodesInstanceBalance(
+                deploymentGroup.getInstances(), ((NodesScalingGroup) scalingGroup).getNodeRefs());
+
+        // Determine where to create each instance
+        int instanceCounter = 0;
+        while (instanceCounter < numberOfNewInstances) {
+            // Get the node with the least instances
+            Map.Entry<String, Integer> maxNodeEntry = Collections.min(
```

Comment:
If it's the nodes with the least instances, why its named `maxNodeEntry`?

Reply:
Copy-pasta error